common/src/main/java/io/netty/util/concurrent/Promise.java

```diff
@@ -59,9 +59,79 @@
      */
     boolean setUncancellable();
 
-    @Override
-    Promise<V> addListener(FutureListener<? super V> listener);
+    /**
+     * Cancel this promise, unless it has already been completed.
+     * <p>
+     * A cancelled promise is considered to be {@linkplain Future#isFailed() failed}.
+     * <p>
+     * If the cancellation was successful it will fail associated futures with a {@link CancellationException}.
+     * <p>
+     * Cancellation will not cause any threads working on the promise to be {@linkplain Thread#interrupt()
+     * interrupted}.
+     *
+     * @return {@code true} if the promise was cancelled by this call, otherwise {@code false}.
+     */
+    boolean cancel();
 
-    @Override
-    <C> Promise<V> addListener(C context, FutureContextListener<? super C, ? super V> listener);
+    /**
+     * Returns {@code true} if and only if the operation was completed successfully.
+     */
+    boolean isSuccess();
```

Comment:
nit: I wonder if we want to add an extra interface for all these methods and have `Future` and `Promise` extend it.

Maybe something like:

```java
public interface AsycResult<V> {
    boolean isSuccess();
    boolean isFailed();
    boolean isCancelled();
    boolean isDone();
    V getNow();
    Throwable cause();
    ....
} 

public Promise<V> extends AsyncResult<V> {
}

public Future<V> extends AsyncResult<V> {
}


```